src/utils/spark.py

```diff
@@ -64,12 +65,37 @@ def __init__(self, app_name: str, password_file_path: str) -> None:
         self.initial_compression = "snappy"
         self.final_compression = "zstd"
 
-        # Store S3 details so workers can upload finished jobs
-        self.s3_client = boto3.client("s3")
-        self.s3_bucket = os.getenv("AWS_S3_BUCKET")
-        self.s3_prefix = os.getenv("AWS_S3_PREFIX", "iasworld")
-
+        # Create the Spark session and logging
         self.spark = SparkSession.builder.appName(self.app_name).getOrCreate()
+        self.log_file_path = f"/tmp/logs/{self.app_name}.log"
+        self.logger = self.get_logger()
+
+    def get_logger(self):
+        """
+        Extract the logger from the JVM used by Spark. We can send Python
+        logging messages here with the same format. Also writes to a file.
+        """
+        # Get Spark logger. See https://stackoverflow.com/a/72740559
+        log4j = self.spark.sparkContext._jvm.org.apache.log4j
+        spark_logger = log4j.LogManager.getLogger("org.apache.spark")
+
+        # Create a file appender to write to write to a session log file
+        appender = log4j.FileAppender()
+        appender.setAppend(True)
+        # For some reason it's necessary to set the log pattern, even though
+        # the format string replicates the default format used by Spark, with
+        # the addition of milliseconds
+        layout = log4j.PatternLayout()
+        layout.setConversionPattern(
+            "%d{yyyy-MM-dd_HH:mm:ss.SSS} %p %c{1}: %m%n"
+        )
+        appender.setLayout(layout)
+        appender.setFile(self.log_file_path)
+        appender.activateOptions()
+
+        spark_logger.removeAllAppenders()
+        spark_logger.addAppender(appender)
```

Comment:
One major annoyance here is that while this setup successfully sinks both Spark and Python log output to the file, only the Spark output is shown in stdout. I couldn't for the life of me figure out how to print both to both file _and_ stdout. That said, I don't think it matters too much since this will be a mostly automatic process and all the logs go to CloudWatch anyways.

Reply:
**[Question, non-blocking]** Now that we're using a Python logger for the non-Spark code, does the script write Python logs to stdout? Or is it still only Spark logs? I don't think it matters either way, I'm just curious!

Reply:
It now writes both Python and Spark logs to both stdout and the log file!